client/mksubvolume.cc

```diff
@@ -326,45 +378,79 @@ doit()
     // Determine name for new subvolume: It is the target name without the
     // leading filesystem target.
 
-    string subvolume_name = target.substr(fs_target.size() + (fs_target == "/" ? 0 : 1));
+    string subvolume_name = target.substr(fs_target.size() +
+					 (fs_target == "/" || fs_target == target ? 0 : 1));
     if (verbose)
 	cout << "subvolume-name:" << subvolume_name << endl;
 
-    // Execute all steps.
+    // Set up cache for UUID / LABEL resolution in mnt_table_find_source
+    libmnt_cache* cache = mnt_new_cache();
+    libmnt_table* mtab_table = mnt_new_table();
+    mnt_table_set_cache(mtab_table, cache);
+    mnt_table_parse_mtab(mtab_table, NULL);
+
+    libmnt_fs* expected_fs = create_fstab_line(fs, subvol_option, subvolume_name);
+
+    // Consistency checks on (partially) existing entries
+    libmnt_fs* fstab_entry = mnt_table.find_target(target, MNT_ITER_FORWARD);
+    libmnt_fs* mounted_entry = mnt_table_find_target(mtab_table, mnt_fs_get_target(expected_fs),
+						     MNT_ITER_BACKWARD);
+    if (subvolume_name.empty())
+	throw runtime_error("target is a dedicated mountpoint");
+    if (fstab_entry != NULL && strcmp(mnt_fs_get_source(fstab_entry), mnt_fs_get_source(expected_fs)) != 0)
+	throw runtime_error("existing fstab entry doesn't match target device");
+    if (fstab_entry != NULL && strcmp(mnt_fs_get_options(fstab_entry), mnt_fs_get_options(expected_fs)) != 0)
+	throw runtime_error("existing fstab entry options don't match");
+    // Something is mounted there already. Is it the correct device?
+    if (mounted_entry != NULL)
+    {
+        char* subvol_expected;
+	size_t valsz;
+	mnt_fs_get_option(expected_fs, "subvol", &subvol_expected, &valsz);
+	// Map UUID / LABEL to a physical device name
+	const char* real_device = mnt_fs_get_source(mnt_table_find_source(mtab_table,
+							mnt_fs_get_source(expected_fs), MNT_ITER_BACKWARD));
+	// Find last device in mtab to get the actual mount
+	mounted_entry = mnt_table_find_target(mtab_table, mnt_fs_get_target(expected_fs),
+					      MNT_ITER_BACKWARD);
+	if (mounted_entry != NULL)
+	{
+	    if (strcmp(real_device, mnt_fs_get_source(mounted_entry)) != 0)
+		throw runtime_error("different device mounted on target");
+	    char* subvol_real;
+	    mnt_fs_get_option(mounted_entry, "subvol", &subvol_real, &valsz);
```

Comment:
Is it OK to not check the return value? What it the subvol entry does not get found?

Reply:
This should not have happened, as we knew at this place that the device contains a btrfs file system, which will always have a `subvol` option. Just to be safe I've added the check, though.